### ISSUE TITLE
Create common UI framework

### ISSUE THREAD

**Issue by Zemledelec** [openglobus/openglobus#561]
I propose to think about how we could improve the UI interface _(based on og control)_ for common purposes. Now, we have `MenuPanel `where we place the buttons, such as `LayerSwitcher`, not exactly placed but in advance `RullerSwitcher`, it could be `DebugDialog `switcher, some timeline interface, etc.

@martyrion proposed `MenuPanel `interface, as a container for the switchers _(buttons)_. I like the idea, but I don't really like how it works now because often we need to use the power of common control outside of the common UI _(switchers)_. So this is a matter of discussion.

The first is to propose a structure _(code design)_ for the elements which will support the interface.

FYI @martyrion @pavletto

**Comment by martyrion** [comment]
Good morning, If I understood correctly we have to find a way to use the controls both **_programmatically_** and by **_user interface_**.
Right now - with the way I suggested - the controls can be activated **ONLY** if user presses a button.
Correct?

**Comment by Zemledelec** [comment]
Hi @martyrion , yes it's right. Check my last commit for `RullerSwitcher`, this is how I think it should be.

**Comment by martyrion** [comment]
Yes, it looks very good and readable. So all people who need to access certain controls via UI, can do it via the corresponding _**switchers**_ . The rest by the control itself. 

Another thought was to have everything related to UI attached to the same object.

```javascript
let ruler = new Ruler (
     control : switcher, // switcher, manual
     ignoreTerrain : false  
)
```
This way we can avoid repeating code and future changes/improvements will be easier. For example maybe someone wants to attach the beginning of a ruler to a fixed `lonlat` and the end to a moving billboard. The object could change to:

```javascript
let ruler = new Ruler (
     control : switcher, // switcher, manual
     ignoreTerrain : false,
     start : [50, 50, 2000],
     end : myFastBillboard
 
)
```

But again, these are just thoughts. I don't have the big picture like you do!

**Comment by Zemledelec** [comment]
I can't agree that `Ruler ` _(as in the example)_ contains a pointer to a switcher. It brings hard links to the design. In my opinion Ruler _(or for example, it could be a drawing control, etc.)_ and other 3D tools are better when they are separate from UI, but UI makes them included for their purposes.

P.S
`start`, `end `in the ruler options - good idea.

**Comment by martyrion** [comment]
Yes, I see what you mean. Good then!

**Comment by Zemledelec** [comment]
All buttons also should be inside a container, `og-menu-bar-vertical` could be the such container.

**Comment by martyrion** [comment]
I will try to implement during this weekend and send some feedback

**Comment by Zemledelec** [comment]
Also, LayerSwitcher must have options to show and hide it manually, and the option to not close it when clicking outside, because it is a useful debug option.

**Comment by Zemledelec** [comment]
Can't figure out, why mouse wheel scrolling doesn't work anymore in the layerswitcher.

**Comment by martyrion** [comment]
I 'll try to look it also

**Comment by martyrion** [comment]
Hey @Zemledelec, I have just noticed that html has some tags, specifically for dialogs and expandable lists. 

[https://css-tricks.com/some-hands-on-with-the-html-dialog-element/](url)
[https://developer.mozilla.org/en-US/docs/Web/HTML/Element/details](url)

Here is the browser compatibility chart:

[https://caniuse.com/details](url) - I think it has good support in almost all browsers since ~ 2016-2020.
[https://caniuse.com/dialog](url) - Supported by modern browsers ~ 2020-2022

If they are used, they will simplify and reduce the code for LayerSwitcher. I think we can safely use details/summary. Regarding dialog tags, the support is still mediocre.

Regarding the requirements of LayerSwitcher:
- Button should be placed inside a menu control.
- Dialog should stay open if we choose to (maybe even place it freely like a floating window).
- Dialog should have an expandable tree structure (again activated with an option):
    -Layer
     ---Entity/Entity Collection
     ------Entity

Do you agree?

**Comment by Zemledelec** [comment]
Absolutely! Very good!

**Comment by Zemledelec** [comment]
Menu betters say a container, not control. Control is like a plugin in terms of openglobus.
Could the dialog be draggable?

**Comment by martyrion** [comment]
Good. Yes that is what I am planning. If it is not dialog-style (which closes when you click outside) then it will be draggable with "X" button, to close

**Comment by martyrion** [comment]
Just to be sure, the following is true:

- A Layer can contain Entities.
- An Entity collection can contain Entities.
- A Layer cannot contain Entity Collections.

I think Entity Collections should live inside a Layer. 

Take for example a scenario that you want to create a drone, with a vertical line to the ground and maybe a camera footprint. Isn't it natural to group them into an Entity Collection, repeat for every other similar case and then put everything to a Layer?

@Zemledelec

**Comment by Zemledelec** [comment]
> A Layer can contain Entities.

Only Vector layer contains entities

Vector layer contains EntityCollections, EntityCollections contains entities, but EntityCollections are not a part of the public (interface) data of a layer. EntityCollections are part of the renderer and help to render(visualize) entities. Vector layer public interface works only with entities. From this point of view, entities become part of a planet, which data represents in vector layers.

You are right! EntityCollections live inside a Vector layer. For different types of entities, there are few EntityCollection containers inside vector layer object. Check it out.

https://github.com/openglobus/openglobus/blob/26ced58ca465aef886fb2a64b93cdf3ca6e80d47/src/og/layer/Vector.js#L129

**Comment by martyrion** [comment]
Hi @Zemledelec,

I have almost completed the new LayerSwitcher.

Functionality:

- Expandable sections
- Entities of a layer are presented and can be selected/unselected
- Dialog can be pinned to a position or unpinned and moved freely
- Dialog can be reset to original position
- Dialog can be minimized/maximized

https://user-images.githubusercontent.com/63229987/190707239-22847c92-9995-4643-ab5e-98aab4c18e10.mov

I would like your thoughts in general and more on the way the entities of a layer are presented. 

Q1: Wouldn't it be cool to have a way to group them? (for example the green area and the respective label)

Q2: Also I have noted that entities -  in this example - do not have a zIndex. Do they generally have? If so I will implement drag-and-drop also for the entities, not only layers.

Q3: Should an entity be allowed to be dragged and dropped to another layer?

Dimitris

**Comment by yeluoyunfei** [comment]
That's very cool. One thoughts from my side, I would prefer that the layers in each group is unfold instead of collapsed, so users can see all the layers directly without extra operation. what do think?

**Comment by martyrion** [comment]
> That's very cool. One thoughts from my side, I would prefer that the layers in each group is unfold instead of collapsed, so users can see all the layers directly without extra operation. what do think?

Hi! If you mean that initial state is unfolded, I think it is good idea. User later can collapse

**Comment by martyrion** [comment]
> That's very cool. One thoughts from my side, I would prefer that the layers in each group is unfold instead of collapsed, so users can see all the layers directly without extra operation. what do think?

Hi! If you mean that initial state is unfolded, I think it is good idea. User later can collapse

**Comment by yeluoyunfei** [comment]
> > That's very cool. One thoughts from my side, I would prefer that the layers in each group is unfold instead of collapsed, so users can see all the layers directly without extra operation. what do think?
> 
> Hi! If you mean that initial state is unfolded, I think it is good idea. User later can collapse

Yes. exactly

**Comment by Zemledelec** [comment]
Hi @martyrion ! It looks impressive! Let me review, please:) 
I agree that unfolded groups as a default state of the interface are more convenient. If you use a component of the group it could be optional.
About your questions about the entities z indexes. 

First of all, I'd like to say that this UI representation makes architecture more obvious! Great job @martyrion! 

Secondly, I suggest do not touch the entities z-indexes right now, because it doesn't work the same way as overlay layers, I suggest (for the future) making some properties available through the UI, both for layers and entities, such as isBaseLayer, zIndex(already draggable), url, name, etc, and for an entity: billboard icon, colors, label text, coordinates, etc. if it is polygon some styles...also each entity has children entities, so it could be nested tree, for the future. 

Thirdly, I don't think that dragging entities is a good idea...right now, but I've never thought about such kind of perspective of using entities around layers. 

In summary, it looks very promising, and reveals some props that would be very helpful, for instance, helps people make their own maps easily! In advance, I think you will like, globus has cameras, and points of interest, and what if it gives creating a demos, where camera fly here and there following a track path...and people can configure this easily!

Cheers!

**Comment by martyrion** [comment]
I am glad you liked it! What do you mean by review? Submit pull request? I will do so in a few days - maybe tomorrow - I want to tidy up the code a bit. I also want you to review the code, I am really interested in your point of view.

So....
- Let's leave entities zIndex for now
- As for the second, I couldn't agree more! That's what I also have in mind. Click on an edit button on a layer/entity and then be able to do everything through the UI, much like a modern application. I think you will be glad to hear that the records of the dialog window are created use recursion, so it doesn't really matter how much nested are items!
- Third --> got that, agree.

have a nice weekend!

**Comment by Zemledelec** [comment]
Yes, review - merge request:)

**Comment by martyrion** [comment]
Hi, what is the way to capture all picking events in a planet?

**Comment by Zemledelec** [comment]
Good!
Hey @martyrion, I've made a few things and merged them with your request.

Let's create another issue, and you can close this.

Btw. properties and methods started with `_`, such as _active are private or protected. You can't use it (actually you can, but better not). 

I didn't get your question about events. Let's discuss it in the forum!
